### PR TITLE
helm api: allow user to specify a version range

### PR DESCRIPTION
**Pull request by natasha41575**
This PR adds support for allowing users to specify a range in `spec.helm.version` in the RSync object and we pull down the latest one from the range.

- The version range syntax is natively supported by Helm CLI. See the syntax here: [https://github.com/Masterminds/semver#hyphen-range-comparisons](https://www.google.com/url?q=https://github.com/Masterminds/semver%23hyphen-range-comparisons) 
- This PR has some changes to ensure that the RSync object is reporting the actual version that was pulled down, instead of just displaying whatever was specified in `spec.helm.version`
- If someone pushes a new helm chart, helm-sync will pick up the new chart without requiring changes to the RSync object.

### CODE REVIEW

File: e2e/testdata/helm-charts/simple-2.5.3/Chart.yaml
Comment (natasha41575) [inline e2e/testdata/helm-charts/simple-2.5.3/Chart.yaml:5]: I don't have permission to push this chart to the `kpt-config-sync-ci-release` project, so I will need someone to do that for me

File: pkg/reconcilermanager/controllers/util.go
Comment (natasha41575) [inline pkg/reconcilermanager/controllers/util.go:99]: If the version is left out entirely, helm CLI will automatically pull the latest one, so there is no need to set a default here. It was discussed offline that supporting polling from a latest tag with helm could be inefficient/not scalable.

File: e2e/testdata/root-sync-helm-chart-version-range-cr.yaml
Comment (natasha41575) [inline e2e/testdata/root-sync-helm-chart-version-range-cr.yaml:26]: This syntax is natively supported by helm CLI.

See [https://github.com/Masterminds/semver#hyphen-range-comparisons](https://www.google.com/url?q=https://github.com/Masterminds/semver%23hyphen-range-comparisons) for more details on the syntax

File: pkg/parse/root.go
Comment (janetkuo) [inline pkg/parse/root.go:248]: Would you make this a function?

File: pkg/reconcilermanager/controllers/util.go
Comment (janetkuo) [reply pkg/reconcilermanager/controllers/util.go:99]: We could also consider keeping it, so that if helm CLI decides to break this default behavior later, we wouldn't be affected.

Comment (natasha41575) [reply pkg/reconcilermanager/controllers/util.go]: Thank you for your comment, as it led me to further investigate this code. I learned that this does not actually affect what gets passed to helm CLI as it populates the `reconciler` container, not the `helm-sync` container. It seems that the primary purpose of this default value is to display `version: latest` in the RSync status when the version is empty (rather than displaying `version: ""`). Since another part of my PR overwrites the RSync status to display the version that was actually pulled down, I think in most cases this bit of code is inconsequential. That said, if something were to go wrong with the version parsing, it would make more sense to display "latest" than "", so I have reverted this change.

File: pkg/parse/root.go
Comment (natasha41575) [reply pkg/parse/root.go]: Good call, done.

File: e2e/testdata/root-sync-helm-chart-version-range-cr.yaml
Comment (janetkuo) [reply e2e/testdata/root-sync-helm-chart-version-range-cr.yaml:26]: Will a new version trigger an update automatically? Is there a way test that? 

FYI - Because the update is triggered without changing R*Sync spec, this is going to make `status.observedGeneration` less useful. Tagging @sdowell given that we discussed this earlier.

**Comment by janetkuo** [comment]
I added a comment https://github.com/GoogleContainerTools/kpt-config-sync/pull/573#discussion_r1190451574 but it's not published as a separate review in the bottom of the page.

File: e2e/testdata/root-sync-helm-chart-version-range-cr.yaml
Comment (natasha41575) [reply e2e/testdata/root-sync-helm-chart-version-range-cr.yaml:26]: > Will a new version trigger an update automatically? Is there a way test that?

I tried it manually and yes, it does an automatic update. 

I'll see if I can add a test. The test would presumably involve pushing a new chart version to the test registry and then deleting it after, which I think is possible, but I will have to look more into it.

Comment (natasha41575) [reply e2e/testdata/root-sync-helm-chart-version-range-cr.yaml:26]: Actually, do you know if the test env has helm CLI on it? I think that would be needed for adding a test

Comment (sdowell) [reply e2e/testdata/root-sync-helm-chart-version-range-cr.yaml:26]: Assuming you are referring to the e2e test env, [yes it has helm CLI](https://github.com/GoogleContainerTools/kpt-config-sync/blob/main/build/test-e2e-go/gke/Dockerfile#L41). Looks like the e2e test runner GSA currently does not have permission to delete images, but that could be added. Why does it need to delete images? For test cleanup?

This would be another use case where using an in-cluster/ephemeral/hermetic registry would be preferred. That shouldn't be a blocker for you here, just wanted to flag it

Comment (natasha41575) [reply e2e/testdata/root-sync-helm-chart-version-range-cr.yaml:26]: > Why does it need to delete images? For test cleanup?

Yes for test cleanup; this test would need to push a new tag and clean up the tag afterwards. Since we can add the permissions, seems like there are no blockers so I will try to add the test.

I agree hermetic registry would be a better solution in the long term.

Comment (natasha41575) [reply e2e/testdata/root-sync-helm-chart-version-range-cr.yaml:26]: Pushed the new test.

/hold (until the e2e test env GSA to have the right permissions)

**Comment by natasha41575** [comment]
/hold 

(see https://github.com/GoogleContainerTools/kpt-config-sync/pull/573#discussion_r1195752922)

File: e2e/testdata/root-sync-helm-chart-version-range-cr.yaml
Comment (janetkuo) [reply e2e/testdata/root-sync-helm-chart-version-range-cr.yaml:26]: @natasha41575 Thanks for adding the test. Would it trigger an upgrade (to the latest version) when a version range is specified, too?

**Review by janetkuo — APPROVED** [review]
/lgtm

**Comment by natasha41575** [comment]
Thanks @janetkuo for the review! I had to make a change to how we push the new helm chart due to the GSA permissions, so will need another LGTM. Also had another question https://github.com/GoogleContainerTools/kpt-config-sync/pull/573#discussion_r1196953251 that we should consider before merging this with the new test

File: e2e/testcases/helm_sync_test.go
Comment (sdowell) [inline e2e/testcases/helm_sync_test.go:231]: Does `gcloud auth configure-docker us-docker.pkg.dev` work? From the [helm](https://cloud.google.com/artifact-registry/docs/helm/authentication#methods)/[gcloud](https://cloud.google.com/artifact-registry/docs/docker/authentication#gcloud-helper) docs it sounds like helm should work with the credential helper.

I think it would be preferred to avoid printing/storing an access token to minimize the risk of leaking a token. Using a credential helper is the most secure option AFAICT

File: e2e/testdata/root-sync-helm-chart-version-range-cr.yaml
Comment (sdowell) [reply e2e/testdata/root-sync-helm-chart-version-range-cr.yaml:26]: Yes, multiple instances of this test may be running concurrently (e.g. for the different GKE channels) - and that's something we need to account for

Comment (natasha41575) [reply e2e/testdata/root-sync-helm-chart-version-range-cr.yaml:26]: @janetkuo What would you like to do here? Should we remove the part of the test I added for automatic updates and merge the PR with just the other tests for now, or should we block on rewriting these tests to be more hermetic/not use an external AR?

Comment (janetkuo) [reply e2e/testdata/root-sync-helm-chart-version-range-cr.yaml:26]: I'd prefer we make the tests more hermetic if possible. It will likely block us from adding more tests in the future.

File: e2e/testcases/helm_sync_test.go
Comment (natasha41575) [reply e2e/testcases/helm_sync_test.go:231]: I tried `gcloud auth configure-docker us-docker.pkg.dev` and it didn't work. If you have other suggestions or know how to get that working I'm happy to try it out, but so far the PAT is the most secure option that I've been able to get to work (and FWIW they are short-lived).

File: e2e/testdata/root-sync-helm-chart-version-range-cr.yaml
Comment (natasha41575) [reply e2e/testdata/root-sync-helm-chart-version-range-cr.yaml:26]: This will be unblocked by https://github.com/GoogleContainerTools/kpt-config-sync/pull/636. When that is merged I will rebase this PR

File: e2e/testcases/helm_sync_test.go
Comment (sdowell) [inline e2e/testcases/helm_sync_test.go:211]: Could this test use workload identity instead (with `e2e-test-ar-reader`)?

Doesn't seem like we're specifically testing token auth here, and I think it would be preferred to use WI as its easier to manage. The token auth with the exported key is more of an edge case IMO and is harder to manage. I'd suggest we stick to just using the exported token for tests which are specifically verifying token auth.

File: e2e/testcases/helm_sync_test.go
Comment (sdowell) [inline e2e/testcases/helm_sync_test.go:227]: The assertions with `WatchForAllSyncs`/`Validate` here and below could be simplified to the following (with the updated version in the latter assertion):
```
nt.Watcher.WatchObject(
  kinds.Deployment(), "deploy-default", "simple",
  []testpredicates.Predicate{testpredicates.HasLabel("version", privateSimpleHelmChartVersion)}
)
```
This will wait for the deployment to have the label in the predicate

File: e2e/testcases/helm_sync_test.go
Comment (natasha41575) [reply e2e/testcases/helm_sync_test.go]: Yup makes sense, updated.

File: e2e/testcases/helm_sync_test.go
Comment (natasha41575) [reply e2e/testcases/helm_sync_test.go]: Thanks for the pointer! The test looks a lot simpler now.

File: e2e/testcases/helm_sync_test.go
Comment (sdowell) [reply e2e/testcases/helm_sync_test.go]: it looks like the test is still using token auth

Comment (natasha41575) [reply e2e/testcases/helm_sync_test.go]: Are you possibly looking at the first commit? I changed auth type in the latest one https://github.com/GoogleContainerTools/kpt-config-sync/pull/573/commits/96a6f44b35f2ab008d1775473d66647d21232d11

File: e2e/testcases/helm_sync_test.go
Comment (sdowell) [inline e2e/testcases/helm_sync_test.go:255]: Does this test not need to check a version label on the deployment?

File: e2e/testcases/helm_sync_test.go
Comment (sdowell) [reply e2e/testcases/helm_sync_test.go]: Ah you're right looks like I was looking at a stale commit, looks good now!

File: e2e/testcases/helm_sync_test.go
Comment (sdowell) [reply e2e/testcases/helm_sync_test.go]: No problem! Agreed looks good

**Review by sdowell — COMMENTED** [review]
Final comments, mostly LGTM

File: e2e/testcases/helm_sync_test.go
Comment (natasha41575) [reply e2e/testcases/helm_sync_test.go:255]: This is a public helm chart made by bitnami, so no, the deployment template does not have a version label.

The other test is a chart one we wrote ourselves, and I added a version label to it myself in this PR for the automatic update test, so that I could check that the new version did indeed get resynced to the cluster. 

This test however is just testing that we grab the latest version from a version range, so checking that the correct version was pulled down is sufficient here, which we do by checking the commit in L249.

Comment (sdowell) [reply e2e/testcases/helm_sync_test.go:255]: Ack, thank you for clarifying

**Review by sdowell — APPROVED** [review]

**Pull request merged**
